### PR TITLE
fix(@aws-amplify/pubsub): correctly match # and + wildcards in message topic

### PR DESCRIPTION
**Pull request by Pwntus**
*Issue #1998, #993*
**Partially** fixes the problem.

*Description of changes:*
Iterate observer topics and normalize topic with incoming message topic for comparison. Observer topic `/foo/#` will match any incoming topic beginning with `/foo/`.

This is most likely an ugly solution, and I don't know about the side-effects it may cause, but I want to initiate a complete fix for this issue.

The + (plus) wildcard is yet to be supported.

More info in MQTT topics & best practices: https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by Pwntus** [comment]
**Update:** Both # and + wildcards are now supported using a `mqttTopicMath()` function traversing each "level" of the filter and message topic to find a mismatch as fast as possible.

File: packages/pubsub/__tests__/PubSub-unit-test.ts [hunk 1]
```diff
@@ -72,6 +72,41 @@ describe('PubSub', () => {
...
+            expect(awsIotProvider.getCategory()).toBe('PubSub');
+
+            const expectedData = {
+                value: 'my message',
+                provider: awsIotProvider
+            };
+            const expectedObserver = {
+                next: data => {
+                    expect(data).toEqual(expectedData);
+                    done()},
```
Comment (elorzafe) [inline packages/pubsub/__tests__/PubSub-unit-test.ts]: You are only testing the first message that arrives. It should have a counter or a mechanism that find the expected messages to be received and then call done on observer.next.

Comment (elorzafe) [reply packages/pubsub/__tests__/PubSub-unit-test.ts]: I think you cannot use `async` and `done()` on the same test. You should use one or the other.

File: packages/pubsub/src/Providers/MqttOverWSProvider.ts [hunk 2]
```diff
@@ -127,14 +127,35 @@ export class MqttOverWSProvider extends AbstractPubSubProvider {
 
     private _onMessage(topic: string, msg: any) {
         try {
-            const observersForTopic = this._topicObservers.get(topic) || new Set();
+            const mqttTopicMath = (filter: string, topic: string) => {
```
[the remark this answers is not in the document]
Comment (Pwntus) [reply packages/pubsub/src/Providers/MqttOverWSProvider.ts]: Thank you for the review! :smiley:
I've added the last test case which shouldn't match on the published topic.

File: packages/pubsub/src/Providers/MqttOverWSProvider.ts [hunk 3]
```diff
@@ -127,14 +127,35 @@ export class MqttOverWSProvider extends AbstractPubSubProvider {
...
+                const filterArray = filter.split('/');
+                const length = filterArray.length;
+                const topicArray = topic.split('/');
+
+                for (let i = 0; i < length; ++i) {
+                    const left = filterArray[i];
+                    const right = topicArray[i];
+                    if (left === '#') return topicArray.length >= length - 1;
```
[the remark this answers is not in the document]
Comment (Pwntus) [reply packages/pubsub/src/Providers/MqttOverWSProvider.ts]: Good catch. Changed according to your suggestion.

File: packages/pubsub/__tests__/PubSub-unit-test.ts (hunk 1 above)
Comment (Pwntus) [reply packages/pubsub/__tests__/PubSub-unit-test.ts]: I removed async and added a `cbCounter` keeping track of each callback, only calling `done()` when the last callback has finished.

**Comment by Pwntus** [comment]
I still think there may be a flaw in the test. The introduced `cbCounter` defeat the purpose of `expect()`. It is impossible to overshoot expected number of assertions.

**Comment by Pwntus** [comment]
`mqttTopicMatch()` is now a public method on class `MqttOverWSProvider`. The test is changed to test this method only.

Additional note: I have tested this in a local development application and it is working as intended.

File: packages/pubsub/src/Providers/MqttOverWSProvider.ts [hunk 4]
```diff
@@ -125,16 +125,37 @@ export class MqttOverWSProvider extends AbstractPubSubProvider {
 
     protected _topicObservers: Map<string, Set<ZenObservable.SubscriptionObserver<any>>> = new Map();
 
+    public mqttTopicMatch(filter: string, topic: string) {
```
Comment (elorzafe) [inline packages/pubsub/src/Providers/MqttOverWSProvider.ts]:
```suggestion
    export function mqttTopicMatch(filter: string, topic: string) {
```
Sorry about that, when I saw the unit test it doesn't make sense for `mqttTopicMatch` to be a method of the class. Is just an static function that needs to be exported (it doesn't depend of MqttProvider). You will need to move this function outside of the class.

File: packages/pubsub/src/Providers/MqttOverWSProvider.ts [hunk 5]
```diff
@@ -125,16 +125,37 @@ export class MqttOverWSProvider extends AbstractPubSubProvider {
...
+        }
+        return length === topicArray.length;
+    }
+
     private _onMessage(topic: string, msg: any) {
         try {
-            const observersForTopic = this._topicObservers.get(topic) || new Set();
+            const matchedTopicObservers = [];
+            this._topicObservers.forEach((observerForTopic, observerTopic) => {
+                if (this.mqttTopicMatch(observerTopic, topic)) {
```
Comment (elorzafe) [inline packages/pubsub/src/Providers/MqttOverWSProvider.ts]:
```suggestion
                if (mqttTopicMatch(observerTopic, topic)) {
```
 When you move `mqttTopicMatch` function outside of the class you will need to change that

File: packages/pubsub/__tests__/PubSub-unit-test.ts [hunk 6]
```diff
@@ -72,6 +72,18 @@ describe('PubSub', () => {
 
             await pubsub.publish('topicA', 'my message');
         });
+
+        test('subscriber is matching MQTT topic wildcards', () => {
+            expect.assertions(5);
+
+            const publishTopic = 'topic/A/B/C';
+            const awsIotProvider = new AWSIoTProvider();
+            expect(awsIotProvider.mqttTopicMatch('topic/A/B/C', publishTopic)).toBe(true);
```
Comment (elorzafe) [inline packages/pubsub/__tests__/PubSub-unit-test.ts]:
```suggestion
            expect(mqttTopicMatch('topic/A/B/C', publishTopic)).toBe(true);
```
After exporting `mqttTopicMatch` function from `MqttOverWSProvider.js`

You will need to import the function and then just  use it on all expect

**Pull request merged**